src/libsystemd/sd-journal/journal-file.c

```diff
@@ -1605,18 +1577,190 @@ static int journal_file_append_data(
         return 0;
 }
 
-uint64_t journal_file_entry_n_items(Object *o) {
-        uint64_t sz;
-        assert(o);
+static int maybe_decompress_payload(
+                JournalFile *f,
+                uint8_t *payload,
+                uint64_t size,
+                int compression,
+                const char *field,
+                size_t field_length,
+                size_t data_threshold,
+                void **ret_data,
+                size_t *ret_size) {
+
+        bool found;
+        int r;
 
-        if (o->object.type != OBJECT_ENTRY)
-                return 0;
+        /* We can't read objects larger than 4G on a 32bit machine */
+        if ((uint64_t) (size_t) size != size)
+                return -E2BIG;
 
-        sz = le64toh(READ_NOW(o->object.size));
+        if (compression != 0) {
+#if HAVE_COMPRESSION
+                if (field) {
+                        r = decompress_startswith(
+                                compression, payload, size, &f->compress_buffer, field, field_length, '=');
+                        if (r < 0)
+                                return log_debug_errno(
+                                        r,
+                                        "Cannot decompress %s object of length %" PRIu64 ": %m",
+                                        object_compressed_to_string(compression),
+                                        size);
+
+                        found = r > 0;
+                } else
+                        found = true;
+
+                if (found) {
+                        size_t rsize = 0;
+
+                        r = decompress_blob(compression, payload, size, &f->compress_buffer, &rsize, 0);
+                        if (r < 0)
+                                return r;
+
+                        if (ret_data)
+                                *ret_data = f->compress_buffer;
+                        if (ret_size)
+                                *ret_size = rsize;
+                }
+#else
+                return -EPROTONOSUPPORT;
+#endif
+        } else {
+                if (field)
+                        found = size >= field_length+1 &&
+                           memcmp(payload, field, field_length) == 0 &&
+                           payload[field_length] == '=';
```

Comment:
Also here, I think the code become simpler by returning earlier if one of these conditions is false.